### PR TITLE
ENH asv benchmarks for Kymatio v0.2+

### LINKED ISSUE #653: asv
https://www.kymat.io/developerguide.html#checking-speed

doesn't work

### PR DESCRIPTION
**Pull request by lostanlen**
Fix #653 
- [X] Enhance the documentation
- [X] Improve the support of `asv` on various frontends, open the possibility to deploy it on GH actions
- [X] Script works for torch (cuda, skcuda, cpu) and numpy (cpu) - it is straightforward to modify the initial script to add more frontends

Note that the batch used when compute the GPU is multiplied by a factor 8: it implies that while benchmarking the torch scat 2D, one uses the exact same hyper param as a batch on iamgenet (can be useful)

> Fixes #653
> So far, only relies on bare-bones dependencies (numpy/scipy)
> @edouardoyallon can you tell me if this runs as intended on your machine? 
> 
> Shall i include Torch and TensorFlow benchmarks in a separate PR? or as a follow-up of this one

### CODE REVIEW

**Comment by eickenberg** [comment]
How does this PR related to the other one relating to asv?

**Comment by lostanlen** [comment]
i had a conversation with @edouardoyallon after he posted #654 and i volunteered to take it from there with a new PR. 
So it supersedes #654

**Comment by eickenberg** [comment]
Thanks, cool.

**Comment by edouardoyallon** [comment]
@lostanlen I propose the following:
- ASV should by default point on the HEAD:master branch of this *online* repo.
- Add in the doc the command line to try ASV on a specific commit.

Once this is done, I'm in favor of merging. @janden / @eickenberg  agreed?

**Comment by edouardoyallon** [comment]
@janden @eickenberg can be merged! feedbacks appreciated. Some adjustements will have to be done when integrating to githubactions (e.g., to which commit we begin to benchmark, etc) + likely the visualization but this is beyond the scope of this PR.

**Comment by edouardoyallon** [comment]
I take it back: asv publish sin't working yet..

**Comment by janden** [comment]
@edouardoyallon Can you summarize a bit what's going on here? For example, why do we have directories named `numpy_` and `torch_`? Looks odd.

**Comment by edouardoyallon** [comment]
Maybe it could be merged?

**Comment by janden** [comment]
Thanks. Will review this afternoon. Rebased on dev to make sure everything is up to date.

**Comment by janden** [comment]
By the way, should this merge into master or into dev?

File: benchmarks/asv_numpy.conf.json
Comment (janden) [inline benchmarks/asv_numpy.conf.json:19]: Can we remove all these commented-out lines? I don't see why we need them.

File: benchmarks/asv_numpy.conf.json
Comment (janden) [inline benchmarks/asv_numpy.conf.json:34]: Same here (and below).

File: benchmarks/asv_numpy.conf.json
Comment (janden) [inline benchmarks/asv_numpy.conf.json:56]: Again, remove commented-out stuff. It's fine to have comments, but they should be actual comments (that may contain code), not just code.

File: benchmarks/asv_numpy.conf.json
Comment (janden) [inline benchmarks/asv_numpy.conf.json:108]: Why underscore?

File: benchmarks/asv_numpy.conf.json
Comment (janden) [inline benchmarks/asv_numpy.conf.json:16]: Do we need all these comments from the template file? Just seems to add noise.

File: benchmarks/asv_torch.conf.json
Comment (janden) [inline benchmarks/asv_torch.conf.json:32]: `master` no (like above)?

File: benchmarks/benchmarks/numpy_/scattering1d.py
Comment (janden) [inline benchmarks/benchmarks/numpy_/scattering1d.py:49]: Why? Aren't all our benchmark numbers going to be twice what they should be?

File: benchmarks/benchmarks/numpy_/scattering3d.py
Comment (janden) [inline benchmarks/benchmarks/numpy_/scattering3d.py:41]: Newline.

File: benchmarks/benchmarks/torch_/scattering1d.py
Comment (janden) [inline benchmarks/benchmarks/torch_/scattering1d.py:68]: I don't like this. Is there no way to include this in the parameters so it's obvious that the batch size is different? Alternatively, would it be so terrible if the CPU takes a long time? Maybe there's a compromise where the CPU is not too slow and the GPU is not too fast?

File: benchmarks/benchmarks/torch_/scattering1d.py
Comment (janden) [inline benchmarks/benchmarks/torch_/scattering1d.py:93]: Newline.

File: benchmarks/benchmarks/torch_/scattering1d.py
Comment (janden) [inline benchmarks/benchmarks/torch_/scattering1d.py:89]: Eh what?

File: benchmarks/benchmarks/torch_/scattering2d.py
Comment (janden) [inline benchmarks/benchmarks/torch_/scattering2d.py:91]: Remove blank.

File: doc/source/developerguide.rst
Comment (janden) [inline doc/source/developerguide.rst:55]: fun = run?

File: doc/source/developerguide.rst
Comment (janden) [inline doc/source/developerguide.rst:67]: No e.g.

File: doc/source/developerguide.rst
Comment (janden) [inline doc/source/developerguide.rst:77]: Right parenthesis after `pytorch`.

File: kymatio/version.py
Comment (janden) [inline kymatio/version.py:2]: Sneaking this in here, huh?

**Review by janden — CHANGES_REQUESTED** [review]

**Comment by edouardoyallon** [comment]
> By the way, should this merge into master or into dev?

I propose to merge this in dev as it brings many new features.

File: benchmarks/asv_numpy.conf.json
Comment (edouardoyallon) [reply benchmarks/asv_numpy.conf.json:19]: done (normally, this config file is automatically created and those comments are still inserted)

File: benchmarks/asv_numpy.conf.json
Comment (edouardoyallon) [reply benchmarks/asv_numpy.conf.json:34]: ok

File: benchmarks/asv_numpy.conf.json
Comment (edouardoyallon) [reply benchmarks/asv_numpy.conf.json:56]: ok

File: benchmarks/asv_numpy.conf.json
Comment (edouardoyallon) [reply benchmarks/asv_numpy.conf.json:108]: because then when you import numpy, "it"(guess who) thinks this is numpy ..

File: benchmarks/asv_torch.conf.json
Comment (edouardoyallon) [reply benchmarks/asv_torch.conf.json:32]: it should be commented out everywhere AFAIU

File: benchmarks/benchmarks/numpy_/scattering1d.py
Comment (edouardoyallon) [reply benchmarks/benchmarks/numpy_/scattering1d.py:49]: I've addressed this so that there is no more those "ifcuda" hidden - i've chosen to pick 2 iteration at least to avoid border effects and reduce the variance of the timing

File: benchmarks/benchmarks/numpy_/scattering3d.py
Comment (edouardoyallon) [reply benchmarks/benchmarks/numpy_/scattering3d.py:41]: done

File: benchmarks/benchmarks/torch_/scattering1d.py
Comment (edouardoyallon) [reply benchmarks/benchmarks/torch_/scattering1d.py:68]: yeah i've removed it. Yes it's terrible if the CPU takes long when you're developping

File: benchmarks/benchmarks/torch_/scattering1d.py
Comment (edouardoyallon) [reply benchmarks/benchmarks/torch_/scattering1d.py:93]: done

File: benchmarks/benchmarks/torch_/scattering1d.py
Comment (edouardoyallon) [reply benchmarks/benchmarks/torch_/scattering1d.py:89]: yeah it's clearer now

File: benchmarks/benchmarks/torch_/scattering2d.py
Comment (edouardoyallon) [reply benchmarks/benchmarks/torch_/scattering2d.py:91]: i've added it

File: doc/source/developerguide.rst
Comment (edouardoyallon) [reply doc/source/developerguide.rst:55]: apple typo correction sorry..

File: doc/source/developerguide.rst
Comment (edouardoyallon) [reply doc/source/developerguide.rst:67]: fixed

File: doc/source/developerguide.rst
Comment (edouardoyallon) [reply doc/source/developerguide.rst:77]: done

File: kymatio/version.py
Comment (edouardoyallon) [reply kymatio/version.py:2]: oO no idea it was done there, sorry

**Comment by edouardoyallon** [comment]
I think I've slightly simplified the mechanics. So, in my opinion we should proceed as follow:
1. Merge this branch (once the code is clear)
2. Propose a PR to run a github action that run ASV (at least on bourreau) and collects the results in some folders
3. Propose a PR that will update the documentation with the collected results.

I can't guarantee that the current PR will work as it is for 2/3 now (because they will sort of correspond to the unit tests to make sure it passes!), but I still think it's fine to merge now, if the code is clean enough.

File: benchmarks/asv_numpy.conf.json
Comment (eickenberg) [reply benchmarks/asv_numpy.conf.json:48]: can you rename them into something else, because this is confusing.

For example `numpy_benchmark`

File: benchmarks/asv_torch.conf.json
Comment (eickenberg) [inline benchmarks/asv_torch.conf.json:51]: `torch_benchmark`

File: benchmarks/asv_torch_cuda.conf.json
Comment (eickenberg) [inline benchmarks/asv_torch_cuda.conf.json:54]: `torch_benchmark`

File: benchmarks/benchmarks/numpy_/scattering1d.py
Comment (eickenberg) [inline benchmarks/benchmarks/numpy_/scattering1d.py:39]:
```python
rng = np.random.RandomState(0)
x = rng.randn(...)
```

File: benchmarks/benchmarks/numpy_/scattering2d.py
Comment (eickenberg) [inline benchmarks/benchmarks/numpy_/scattering2d.py:36]:
```python
rng = np.random.RandomState(0)
x = rng.randn(...)
```

File: benchmarks/benchmarks/numpy_/scattering3d.py
Comment (eickenberg) [inline benchmarks/benchmarks/numpy_/scattering3d.py:29]:
```python
rng = np.random.RandomState(0)
x = rng.randn(...)
```

File: benchmarks/benchmarks/torch_/scattering1d.py
Comment (eickenberg) [inline benchmarks/benchmarks/torch_/scattering1d.py:76]:
```python
rng = np.random.RandomState(0)
x = rng.randn(...)
```

File: benchmarks/benchmarks/torch_/scattering2d.py
Comment (eickenberg) [inline benchmarks/benchmarks/torch_/scattering2d.py:72]:
```python
rng = np.random.RandomState(0)
x = rng.randn(...)
```

File: benchmarks/benchmarks/torch_/scattering3d.py
Comment (eickenberg) [inline benchmarks/benchmarks/torch_/scattering3d.py:64]:
```python
rng = np.random.RandomState(0)
x = rng.randn(...)
```

**Review by eickenberg — APPROVED** [review]
I put some comments for improving the specific code proposed here.

I am OK with merging this for time reasons, but I would prefer to see the following modifications:

- benchmarks for TF
- there is a lot of redundancy in the code that could quite easily be reduced by importing common stuff from a common file and modifying only with specifics from the specific benchmark.

**Comment by edouardoyallon** [comment]
> I put some comments for improving the specific code proposed here.
> 
> I am OK with merging this for time reasons, but I would prefer to see the following modifications:
> 
> * benchmarks for TF
> * there is a lot of redundancy in the code that could quite easily be reduced by importing common stuff from a common file and modifying only with specifics from the specific benchmark.

This will be in a future work if it's ok

**Review by edouardoyallon — APPROVED** [review]

**Comment by edouardoyallon** [comment]
Screen Shot 2021-03-18 at 09 57 26
so you both request changes and didn't review. Interesting

**Comment by edouardoyallon** [comment]
oh sorry I thought you accepted the PR and this was a bug. I'm going to reverse this commit and re-open it

**Pull request merged**
